client/ayon_core/pipeline/create/creator_plugins.py

```diff
@@ -831,6 +833,73 @@ def get_pre_create_attr_defs(self):
         """
         return self.pre_create_attr_defs
 
+    def apply_staging_dir(self, instance):
+        """Apply staging dir with persistence to instance's transient data.
+
+        Method is called on instance creation and on instance update.
+
+        Args:
+            instance (CreatedInstance): Instance for which should be staging
+                dir applied.
+
+        Returns:
+            Optional[str]: Staging dir path or None if not applied.
+        """
+        create_ctx = self.create_context
+        product_name = instance.get("productName")
+        product_type = instance.get("productType")
+        folder_path = instance.get("folderPath")
+
+        # this can only work if product name and folder path are available
+        if not product_name or not folder_path:
+            return
+
+        version = instance.get("version")
+        if version is not None:
+            template_data = {"version": version}
+        else:
+            template_data = {}
+
+        # TODO: confirm feature
+        publish_settings = self.project_settings["core"]["publish"]
+        anatomy_settings = publish_settings["CollectAnatomyInstanceData"]
+        follow_workfile_version = anatomy_settings["follow_workfile_version"]
+        if follow_workfile_version:
+            current_workfile = self.create_context.get_current_workfile_path()
+            workfile_version = get_version_from_path(current_workfile)
+            template_data = {"version": int(workfile_version)}
```

Comment:
@jakubjezek001 @iLLiCiTiT how do you guys like the implementation above.

Basically this make `{version}` in the staging directory regex only available when `ayon+settings://core/publish/CollectAnatomyInstanceData/follow_workfile_version` is enabled.

This way, when enabled `version in workfile == version in staging directory == version of the published product(s)` _see green_.

Otherwise, it does not fail but does not format it. Meaning each instance would override a common `/foo/bar_{version}` directory _see orange_.

![image](https://github.com/user-attachments/assets/13f73c4a-40ee-4fef-a6de-4f8d175a23e6)

Reply:
I can't tell if the workflow concept is ok, @jakubjezek001 can answer that.

From code point of view I think we should always fill it.
```python
        workfile_version = None
        if follow_workfile_version:
            current_workfile = self.create_context.get_current_workfile_path()
            try:
                workfile_version = get_version_from_path(current_workfile) 
            except IndexError:
                pass

        version = "versionless"  # Made this up, please change
        if workfile_version is not None:
            version = int(workfile_version)
        
        template_data = {"version": version}
```

Reply:
The logic is right - yes it should be looking into the settings `ayon+settings://core/publish/CollectAnatomyInstanceData/follow_workfile_version` to check if it should follow workfile version. But if it is not following workfile version it should find next available version for the product. The function is implemented here https://vscode.dev/github/ynput/ayon-core/blob/feature/AY-971_Use-custom-staging-dir-functions/client/ayon_core/pipeline/create/creator_plugins.py#L625